articles/2019-03-25-le-type-option-c-est-quoi-et-ca-regle-quel-probleme.md

```diff
@@ -0,0 +1,177 @@
+
+Cet code va retourner:
+
+- un objet s'il a un champ `active` ayant une valeur évaluée comme vraie
+- `undefined` si un item de `array` est `undefined`
+- `undefined` si rien n'est trouvé
+
+Avec cette implémentation naïve, nous sommes incapables de savoir dans quel cas nous nous trouvons : soit on a trouvé un item `undefined`, soit on a rien trouvé.
+
+_Notez que le problème se pose ici avec `undefined` mais qu'il en serait de même un tableau contenant des `null` et une fonction `find` d'une bibliothèque retournant `null` dans le cas où elle ne trouve rien_
+
+Si l'on veut être capable de faire la différence entre les deux derniers cas, on doit utiliser une autre fonction: `findIndex`:
+
+```js
+let index = array.findIndex(item => item === undefined || item.active);
+if (index == -1) {
+  // not found
+} else {
+  // found
+  let result = array[index];
+}
+```
+
+Le code est plus lourd, moins lisible, et manque d'expressivité. `find` ne nous donne pas assez d'information au travers de la valeur retournée: `undefined` est "applati", et requiert une logique supplémentaire (ici `index`, si un item est trouvé, il sera supérieur à `-1`, sinon il sera égal à `-1`)
+
+Le problème ne vient pas de la fonction `find` elle même mais de la façon dont `null` et `undefined` sont traités. `null` **est** la valeur, il la **remplace**. `option` **l'enrobe**: c'est un conteneur.
+
+```reason
+/* `getBy` est l'equivalent de `find` */
+let result = array->Belt.Array.getBy(
+  fun
+    | None => true
+    | Some({active}) => active
+);
+```
+
+D'abord, `array` a le type suivant:
+
+```reason
+let array: array(option(value));
+```
+
+Et `getBy` celui-ci:
+
+```reason
+let getBy: (array('a), 'a => bool) => option('a);
+```
+
+Si on remplace les paramètres de type par le type vraiment utilisé dans notre cas précis, on se retrouve avec ça :
+
+```reason
+let getBy:
+  (
+    array(option(value)),
+    option(value) => bool
+  ) => option(option(value));
+```
+
+`result` aura donc le type suivant :
+
+```reason
+let result: option(option(value));
+```
+
+C'est une `option` d'`option` de `value`. Et ça signifie qu' **on peut extraire l'information qui nous intéresse** de la valeur de retour:
+
+- si le résultat est `Some(Some(value))` : on a trouvé une valeur `true` pour le champ `active`
+- si le résultat est `Some(None)` : on a trouvé une valeur `None`
+- si le résultat est `None` : on n'a rien trouvé dans le tableau
+
+Le type `option` a éliminé par _design_ certains problèmes inhérents à `null` et `undefined` en se comportant comme un conteneur plutôt qu'un substitut.
+
+La nature des `option` dans les langages statiquement typés permet d'éviter de nombreuses erreurs de conception. Les fonctions ne sont plus juste autorisées à retourner `null` et à vous laisser la responsabilité implicite de le gérer à grand coup de `if(value == null) { a } else { b }`, elles retournent un type `option` qui vous **force à prendre en compte l'optionalité de la valeur**.
+
+Bisous.
```

Comment:
Je trouve que le "bisous" (que je trouve souvent TOUJOURS bien placé) fait un peu jeté là.
Je mettrais un truc plus en suivant genre "C'est pas beau?" ou une formule un peu bateau pour "sortir".

Reply:
resolved in dc09535